### ISSUE TITLE
Linker errors when forgetting `#[no_link]` on a plugin

### ISSUE THREAD

**Issue by tomjakubowski** [rust-lang/cargo#1155]
Not sure if this is a Cargo or rustc issue, but my first inclination is to think that Cargo is passing some wrong paths or something.

I've been debugging this linker error in the `json_macros` crate. I narrowed the reproduction down to this: https://github.com/tomjakubowski/json_macros/tree/repro

Here are the errors on `cargo clean && cargo run --verbose --example kitchen-sink`:

http://sprunge.us/UTCT

If I add `#[no_link]` to the `extern crate json_macros;` in the example, it works just as expected. It looks to me like the example and `json_macros` are somehow linking to different versions of the `rustc-serialize` crate, but I'm not sure.

```
[tom@hannibal json_macros]$ cargo --version
cargo 0.0.1-pre (1a1868b 2014-12-31 21:39:41 +0000)
```

**Comment by tomjakubowski** [comment]
Ah, I think this will be just fixed by rust-lang/rust#21043

**Comment by alexcrichton** [comment]
Yeah unfortunately plugins currently require `#[no_link]` to be placed on them and aren't too useful otherwise! I'm going to close this in favor of https://github.com/rust-lang/rust/issues/21043 as it will likely be smoothed over with that.
